packages/abi/src/parser/specifications/v1/resolvable-type.ts

```diff
@@ -0,0 +1,269 @@
+    abi: AbiSpecificationV1,
+    metadataType: AbiMetadataTypeV1,
+    args: (ResolvableType | ResolvedType)[] | undefined
+  ): Array<[number, ResolvedType | ResolvableType]> | undefined {
+    if (!args) {
+      return metadataType.typeParameters?.map((typeParameter) => [
+        typeParameter,
+        new ResolvableType(abi, typeParameter, undefined),
+      ]);
+    }
+
+    return metadataType.typeParameters?.map((typeParameter, idx) => [typeParameter, args[idx]]);
+  }
+
+  private static handleComponent(
+    parent: ResolvableType,
+    abi: AbiSpecificationV1,
+    c: AbiComponentV1 | AbiTypeArgumentV1
+  ): ResolvableComponent {
+    const name = (c as AbiComponentV1).name;
+
+    if (typeof c.typeId === 'string') {
+      const concreteType = abi.concreteTypes.find(
+        (ct) => ct.concreteTypeId === c.typeId
+      ) as AbiConcreteTypeV1;
+      return {
+        name,
+        type: ResolvableType.resolveConcreteType(abi, concreteType),
+      };
+    }
+
+    const mt = abi.metadataTypes.find((tm) => tm.metadataTypeId === c.typeId) as AbiMetadataTypeV1;
+
+    return {
+      name,
+      type: ResolvableType.handleMetadataType(parent, abi, mt, c.typeArguments),
+    };
+  }
+
+  /**
+   * Concrete types are *resolved* because everything is known about them.
+   */
+  private static resolveConcreteType(
+    abi: AbiSpecificationV1,
+    type: AbiConcreteTypeV1
+  ): ResolvedType {
+    if (type.metadataTypeId === undefined) {
+      return new ResolvedType({
+        type: type.type,
+        typeId: type.concreteTypeId,
+      });
+    }
+
+    if (!type.typeArguments) {
+      return new ResolvableType(abi, type.metadataTypeId, undefined).resolveInternal(
+        type.concreteTypeId,
+        undefined
+      );
+    }
+
+    const metadataType = abi.metadataTypes.find(
+      (mt) => mt.metadataTypeId === type.metadataTypeId
+    ) as AbiMetadataTypeV1;
+
+    const concreteTypeArgs = type.typeArguments.map((ta) => {
+      const concreteTypeArg = abi.concreteTypes.find(
+        (ct) => ct.concreteTypeId === ta
+      ) as AbiConcreteTypeV1;
+      return ResolvableType.resolveConcreteType(abi, concreteTypeArg);
+    });
+
+    return new ResolvableType(
+      abi,
+      type.metadataTypeId,
+      ResolvableType.mapTypeParametersAndArgs(abi, metadataType, concreteTypeArgs)
+    ).resolveInternal(type.concreteTypeId, undefined);
+  }
+
+  /**
+   * Metadata types are *handled* and not *resolved* because they might be generic,
+   * in which case they cannot be resolved.
+   * If they're not generic, they can be immediately resolved.
+   */
+  private static handleMetadataType(
+    parent: ResolvableType,
+    abi: AbiSpecificationV1,
+    mt: AbiMetadataTypeV1,
+    typeArguments: AbiComponentV1['typeArguments']
+  ): ResolvableType | ResolvedType {
+    if (swayTypeMatchers.generic(mt.type)) {
+      const resolvedTypeParameter = parent.typeParamsArgsMap?.find(
+        ([tp]) => tp === mt.metadataTypeId
```

Comment:
We should avoid using abbreviated variable names such as `c`, `ct`, `mt`, `tm`, and instead opt for more descriptive names like component, `concreteType`, `metadataType`, and `typeArgId`.

While these abbreviations might be intuitive for those deeply familiar with the codebase, they can make it significantly harder for newcomers to understand and work with the code. Clear and descriptive naming helps improve readability and maintainability for everyone.

Reply:
fe56f5326289f059b568791effba48f31a608bb0